blockwork/common/yamldataclasses.py

```diff
@@ -0,0 +1,218 @@
+# Copyright 2023, Blockwork, github.com/intuity/blockwork
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Generic, Optional, TypeVar
+from pathlib import Path
+import sys
+from dataclasses import _MISSING_TYPE, fields
+import yaml
+try:
+    from yaml import CDumper as Dumper
+    from yaml import CLoader as Loader
+except ImportError:
+    from yaml import Dumper, Loader
+
+
+class YamlDataclassError(yaml.YAMLError):
+    'Error parsing yaml field'
+    def __init__(self, location: str, msg: str):
+        self.location = location
+        self.msg = msg
+
+    def __str__(self):
+        return f"{self.location}: {self.msg}"
+
+
+class YamlFieldError(YamlDataclassError):
+    'Error parsing yaml field'
+    def __init__(self, location: str, ex: Exception, field: Optional[str] = None):
+        self.field = field
+        self.orig_ex = ex
+        field_str = "" if self.field is None else f" at field `{self.field}`"
+        super().__init__(f"{location}{field_str}", str(ex))
+
+
+class YamlMissingFieldsError(YamlDataclassError):
+    def __init__(self, location: str, fields: list[str]):
+        self.fields = fields
+        super().__init__(location, f"Missing field(s) `{', '.join(self.fields)}`")
+
+
+class YamlExtraFieldsError(YamlDataclassError):
+    def __init__(self, location: str, fields: list[str]):
+        self.fields = fields
+        super().__init__(location, f"Got extra field(s) `{', '.join(self.fields)}`")
+
+
+def constructorFactory(dc: type):
```

Comment:
Elsewhere in blockwork's codebase the code is using lowercase snake style (i.e. `xyz_abc`) and not camelcase it would be good to keep it consistent